illumos-utils/src/zfs.rs

```diff
@@ -375,7 +384,52 @@ impl fmt::Display for PropertySource {
     }
 }
 
-#[cfg_attr(any(test, feature = "testing"), mockall::automock, allow(dead_code))]
+#[derive(Copy, Clone, Debug)]
+pub enum WhichDatasets {
+    SelfOnly,
+    SelfAndChildren,
+}
+
+// A helper structure for gathering all possible key/value pairs to pass to "zfs
+// set". This helps callers minimize the number of calls to "zfs set" they need
+// to make.
+struct PropertySetter {
+    props: Vec<(&'static str, String)>,
```

Comment:
Switching this from `BTreeMap` to `Vec` feels like an awkward compromise - if a caller calls `add_size_details` twice, this type allows it, but the resulting list will be rejected by `zfs`. I still think a standalone function would be clearer; we only know we need these two things (`SizeDetails`, `DatasetUuid`) today, it could return a `Vec` with no ambiguity about duplicate keys, and it could handle another few arguments before becoming unwieldy (which might cover us for months to forever - we don't _know_ that we're going to need more here, right?).

If you feel strongly that you want to keep the separate type here, I'd probably switch this back to `BTreeMap`; it's more obvious what the behavior will be if the methods are called multiple times.

Reply:
I don't feel too strongly, and fair enough, the free function has a clear advantage for now. Updated to use it (no duplicate keys, no unnecessary allocations)